### PR TITLE
RFD 0001: Design for initial implementation of the Load Balancer

### PR DESCRIPTION
**Pull request by xdire**
This RFD Describes plan for initial implementation of the Load Balancer along with considerations of features can be added in future releases.

### CODE REVIEW

File: rfd/0001-general-design.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+- Provide balancing with available algorithms
+
+###### Balancing Algorithms
+This design limits initial implementation to have the following algorithms:
+1) Round Robin
+   - implementation in a form of atomic counter to the size of the pool of servers
+2) Least connection
+   - implementation as a range over the pool of server records with atomically incremented counters
+
+#### API Interface
```
Comment (strideynet) [inline rfd/0001-general-design.md]: To keep the scope smaller, we don't require that the load-balancer offers an API for configuration. It should be fine to just have this configuration be static.

File: rfd/0001-general-design.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+
+## What
+
+This RFD Details the aspects of implementation of TCP load balancer
+up the following minimal requirements:
+
+### Library
+- A least connections forwarder implementation that tracks the number of connections per upstream.
+- A per-client connection rate limiter implementation that tracks the number of client connections.
+- Health checking for unhealthy upstreams. Should not be eligible to receive connections until upstream determined to be healthy.
```
Comment (strideynet) [inline rfd/0001-general-design.md:16]: Could you please provide more details on how you will implement a health checking algorithm for the upstreams ?

File: rfd/0001-general-design.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+# RFD 1 - General Design
+
+## What
+
+This RFD Details the aspects of implementation of TCP load balancer
+up the following minimal requirements:
+
+### Library
+- A least connections forwarder implementation that tracks the number of connections per upstream.
+- A per-client connection rate limiter implementation that tracks the number of client connections.
```
Comment (strideynet) [inline rfd/0001-general-design.md:15]: Could you please provide more details on how you will implement a per-client connection rate-limiter?

File: rfd/0001-general-design.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+---
+`GET` `/api/v1/frontend/list` - requires token, gives back list of frontends which belongs to the client
+
+---
+`DELETE` `/api/v1/frontend/{uuid}` - requires token, will remove frontend from the dispatch and storage
+
+---
+
+### Storage
```
Comment (strideynet) [inline rfd/0001-general-design.md]: Again, to limit scope, you shouldn't actually need any form of persistence to complete this challenge.

**Review by strideynet — COMMENTED** [review]
A good start but I'm a little worried that you've suffered significant scope creep. You shouldn't need an API or Persistence to satisfy the challenge and my main concern is that the increase in scope will mean you will run out time.  The inclusion of this scope creep has also meant that the RFD currently lacks details on how you will implement the key requirements outlined in the challenge document.

File: rfd/0001-general-design.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+                                                   └────┬┴─│├─────────────────┤│──┐      ▲         └──────────────┘
+                                                        └──□│   timeoutSec    ││  │      │                         
+                                                           │└─────────────────┘│  └──────┘                         
+                                                           └───────────────────┘                                   
+```
+#### Frontend
+Is the part of the application customer can configure, and it represents Load Balancer
+object options.
+
+Frontend Available in Api Interface, and is entity in Storage Abstraction.
```
Comment (smallinsky) [inline rfd/0001-general-design.md]: What would you feel about cutting the scope and removing fronted and keeping configuration hardcoded  in code  and focusing in LB functionality ?

File: rfd/0001-general-design.md [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+
+#### Backend
```
Comment (smallinsky) [inline rfd/0001-general-design.md]: What if the difference between Backend and LoadBalancer ?

File: rfd/0001-general-design.md [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+the request. Backends have the following functionality:
+- Provide a routine to pipe the traffic
+- Apply routing strategy
+- Manage connection health per pool
+- Provide balancing with available algorithms
+
+###### Balancing Algorithms
+This design limits initial implementation to have the following algorithms:
+1) Round Robin
+   - implementation in a form of atomic counter to the size of the pool of servers
```
Comment (smallinsky) [inline rfd/0001-general-design.md]: LB Requirement for [L5](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-2.md#level-5) doesn't require to implement round robing so please feel to cut the scope.

File: rfd/0001-general-design.md [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+┌──────┐  │ svc3.x.c│ 8081│ A N  │──┴┤                │  ┌──────────┐   │
+│      │  └─┬───▶───┴────┬┘ D C  │   │  ┌─────────┬───┴──┤ Service  │   │
+│Client│────┘            │    E  │   └──▶ .72.93  │ 50051│ Instance │   │
+│      │                 │    R  │      └─────────┴───┬──┴──────────┘   │
+└──────┘                 └───────┘                    └─────────────────┘
+```
+
+### Components
+
+For our implementation, we will consider the following scheme:
```
Comment (smallinsky) [inline rfd/0001-general-design.md]: Trying to understand the barrier between Frontend and Backend components in your design. 

Also it seems like R1 - R2 routing:
``` 
                          ---- R1
8090 │ mTLS 
                        ----- R2
```

is preformed before reaching fronted or backend. Based on what the 8090 │ mTLS  connection will be routed as R1 or R2 to Fronted1 or Fronted2 ?

File: rfd/0001-general-design.md [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+- Apply routing strategy
+- Manage connection health per pool
+- Provide balancing with available algorithms
+
+###### Balancing Algorithms
+This design limits initial implementation to have the following algorithms:
+1) Round Robin
+   - implementation in a form of atomic counter to the size of the pool of servers
+2) Least connection
+   - implementation as a range over the pool of server records with atomically incremented counters
```
Comment (rosstimothy) [inline rfd/0001-general-design.md]: L5 only requires the least connection algorithm. You may drop support for round robin to reduce scope.

File: rfd/0001-general-design.md (hunk 1 above)
Comment (rosstimothy) [reply rfd/0001-general-design.md]: +1. Reduce scope by eliminating the API entirely in favor of static configuration.

File: rfd/0001-general-design.md [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+###### Considerations
+> `/api/v1/client/auth` can experience bruteforce attacks
+- **Mitigation**: Auth API can be Throttled per IP basis to 10 rps, as this API 
+should not provide capabilities for rapid traffic increases
+- **Security**: Both ClientId and ClientKey should have enough entropy in a combination
+- **Stolen credentials**: Client record can be disabled and this should trigger all the 
+Frontend records belonging to the customer change their Access Key which will invalidate
+selector capabilities to route the traffic for those records
+- **DDoS**: As we scale balancer using the centralized storage, and balancer considered to be
+ephemeral in context of multiple-replicas — we can lose replicas and create new replicas on the go
```
Comment (rosstimothy) [inline rfd/0001-general-design.md]: To reduce scope this can be eliminated entirely in favor of using preconfigured self signed certificates.

File: rfd/0001-general-design.md [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+  string   frontendId = 2;
+  Strategy strategy   = 3;
+  repeated BackendRoute routes = 4;
+}
+
+message BackendRoute {
+  string route       = 1;
+  uint32 connections = 2;
+}
+```
```
Comment (rosstimothy) [inline rfd/0001-general-design.md]: There is no requirement to use protocol buffers for this challenge. You may omit to reduce scope.

File: rfd/0001-general-design.md [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+- API Interface interacting with Centralized Database to scale up Frontends
+- Container can receive events from Pub-Sub to activate/deactivate/invalidate certain resource and
+change the behavior
+
+### Test Plan
+
+For the scope of this design tests will be designed in following areas:
+- Tests for routing strategies to ensure proper operation capabilities and proper concurrency implementation
+- Tests that components follow the context closure
+- Functional tests with concurrency to ensure traffic flows without issues between clients and customer pool
```
Comment (rosstimothy) [inline rfd/0001-general-design.md]: You can remove these Teleport specific sections from the design document since they aren't required for the challenge.

**Review by rosstimothy — COMMENTED** [review]
Agree with Noah, there are some good things here, but most of the design document covers features that are not required for the challenge. Please update the document to omit added features and include more detail for how the requirements will be satisfied.

>Level 5
>Library
-Implement a least connections forwarder that tracks the number of connections per upstream.
-Implement a per-client connection rate limiter that tracks the number of client connections.
-Add health checking: unhealthy upstreams should not be eligible to receive connections until they are determined to be healthy.
>Server
-Use mTLS authentication to have the server verify identity of the client and client of the server
-Develop a simple authorization scheme that defines what upstreams are available to which clients; this scheme can be statically defined in code
-Accept and forward connections to upstreams using library 

https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-2.md#level-5

File: rfd/0001-general-design.md [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+                                                      └────────────────────────────────────────┘
+```
+###### Scope 
+For the scope of the project and simplicity of initial implementation we consider following:
+- Encryption Key RSA 3072
+- Cipher Suites: TLS v1.3 compliant set
+- Client Common Name will be pre-filled with Frontend Access Key (for active selectors)
+- TLS v1.3 as default configuration
+###### Considerations
+- mTLS Layer will match CN Access Key to the available Access Keys in the connection manager
```
Comment (smallinsky) [inline rfd/0001-general-design.md]: Does `Access Key` == `Access Token` ?

File: rfd/0001-general-design.md (hunk 6 above)
Comment (xdire) [reply rfd/0001-general-design.md]: As I mentioned, Frontend and Backend are mostly logical components of the application itself to structure the code in a certain way. An example is in POC: https://github.com/xdire/xlb-poc/blob/main/balancer/backend.go which spawned by the https://github.com/xdire/xlb-poc/blob/main/balancer/manager.go links are to visually explain the choice of wording in design.

File: rfd/0001-general-design.md (hunk 8 above)
Comment (xdire) [reply rfd/0001-general-design.md]: As mentioned in the mTLS section, it will be routed based on CN bundled in the Certificate in the connection manager, which consumes the Frontend record as the entry to start routing mechanics. The CN routing is visually explained here in POC: https://github.com/xdire/xlb-poc/blob/main/balancer/manager.go#L143

Based on the property of the Certificate to be bonded to the signature validity.

File: rfd/0001-general-design.md (hunk 13 above)
Comment (xdire) [reply rfd/0001-general-design.md]: `AccessToken` for `API Interface` in form of `Bearer JwtAccessToken`

`AccessKey` is bundled with `Frontend Certificate CN` for mTLS to decide routing based on certificate credentials.

I assume this is a valid routing method for this challenge, so please feel free to suggest another direction.

File: rfd/0001-general-design.md (hunk 1 above)
Comment (xdire) [reply rfd/0001-general-design.md]: Given all of the commentary, the scope needs to be reduced only to the library implementation with few interfaces to ingest the options to adjust the behavior. I will address this today.

File: rfd/0001-general-design.md (hunk 2 above)
Comment (xdire) [reply rfd/0001-general-design.md:16]: This will be provided in the design update.

The current path is to:
- If Dial failed to one of the members of the pool, put it into the HealthCheck queue
- Mark it as unhealthy (prevent selectors to select)
- time.Ticker the `conn, err := net.DialTimeout("tcp", address, 1*time.Second)`
- if successfully resolved, mark healthy, pause the HealthCheck queue to await unhealthy elements

File: rfd/0001-general-design.md (hunk 3 above)
Comment (xdire) [reply rfd/0001-general-design.md:15]: This will be addressed in the design update:

The current path is:
- As mentioned in mTLS section we decide the connection to be accepted/dropped by the decisions of certificate validity and CN=AccessKey for the Frontend record which designates the customer pool
- As there can be multiple Frontends working on the same server, we can implement TokenBucket algorithm for each Frontend UUID to have it's own TokenBucket updated on the next connection through designated port (ports not in the scope of this design)
- Each new connection will subtract Tokens from the bucket and replenish some if time-window was wide enough

File: rfd/0001-general-design.md (hunk 5 above)
Comment (xdire) [reply rfd/0001-general-design.md]: Thanks for the valid points; the scope will be cut to just a library implementation. 

TLDR:
To clarify, Frontend here is a component of a two-part logical separation in the code, representing the API Entity provided from the customer side, but merely an assembly of properties like: `active` `routes,` etc.

The backend here is a component looking into the customer pool and cannot be affected by the customer using the library other than by the customer updating the Frontend record, which will update the Backend behavior.

Code examples are mentioned below.

File: rfd/0001-general-design.md [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+│  CACert  │───────┐        ┌────────────┐         │  │       └───────┬───────┘                │
+│          │       │        │            │         │  │               │                        │
+└──────────┘  ┌────┴──┐     │   Client   │  ┌──────┴┐ │       ┌───────▼───────┐                │
+              │Produce│     │    Cert    ├──┤Connect│ │       │  Fetch CN     │                │
+┌──────────┐  │  TLS  ├────▶│            │  └───────┘ │       └───────────────┘                │
+│  Client  │  └────┬──┘     │ CN=Identity│            │               │                        │
+│ Frontend │       │        │            │            │       ┌───────┘                        │
+│  Access  │───────┘        └────────────┘            │       │                                │
+│    key   │                                         ┌┴───────▼───────┐                        │
+└──────────┘                                         │ Lookup Access  │    ┌─────────────────┐ │
```
Comment (smallinsky) [inline rfd/0001-general-design.md:366]: It loos like the Client Fronted Access Key block is now outdated

**Review by smallinsky — COMMENTED** [review]
One of L5 requirements  [says](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-2.md#server-4): 

> Develop a simple authorization scheme that defines what upstreams are available to which clients; this scheme can be statically defined in code

Could you elaborate in design docs how are you going to meet this requirement and provide some schema example ?

File: rfd/0001-general-design.md (hunk 14 above)
Comment (xdire) [reply rfd/0001-general-design.md:366]: Yes, I was changing the naming for clarity. At least in my head, I was thinking it would bring one :)

But answering the question:
- CN provides identity key
- Identity Key placed in the CN by outer layers of the application importing the library
- Identity Key is the Identity of the Cluster Pool record (can be UUID of customer pool) or token derived from the customer pool

Let me know if my assumption needs to be corrected for this area, as I prefer to use CN for this type of authentication and might be unaware if other fields are better for this particular goal.

File: rfd/0001-general-design.md (hunk 2 above)
Comment (strideynet) [reply rfd/0001-general-design.md:16]: Thanks for clarifying - please include this within the document.

Comment (xdire) [reply rfd/0001-general-design.md:16]: The document has been updated. I am still fixing typos and re-reading before re-requesting the review; however, you can parse through the updated file right away.

File: rfd/0001-general-design.md (hunk 14 above)
Comment (strideynet) [reply rfd/0001-general-design.md:366]: Using the Common Name/Subject seems like a good strategy here and does match a majority of x509 client cert auth implementations in the wild.  I think a little additional explanation on how the mapping/authorization will work here would be useful - how will access be granted to specific identities, are you pursuing a RBAC model or something different (e.g users are directly assigned access to certain applications).

File: rfd/0001-general-design.md [hunk 15]
```diff
@@ -0,0 +1,288 @@
+---
+authors: Anton Repin (github.com/xdire)
+state: wip
+---
+
+# RFD 1 - General Design
```
Comment (strideynet) [inline rfd/0001-general-design.md:6]: Given that you've switched from a dynamic config API to static configuration, it'd be helpful to see an example of the structure of the static configuration whether that's the configuration file shape or the CLI flags that will be available when executing the load-balancer. I find that considering the configuration ahead of time often reveals complexities in the UX of operating a thing.

File: rfd/0001-general-design.md (hunk 14 above)
Comment (xdire) [reply rfd/0001-general-design.md:366]: Thanks, Noah; as I said above, I updated the document but didn't request the review. But let me see if I need to include additional information after your comment above.

There are certainly new Diagrams there showing how it should work, but I will try to re-check if additional details can be included.

File: rfd/0001-general-design.md (hunk 2 above)
Comment (strideynet) [reply rfd/0001-general-design.md:16]: Thanks!

Comment (strideynet) [reply rfd/0001-general-design.md:16]: Perhaps my eyes are missing it - but I can't see a succinct explanation similar to this within the document - it might've been lost in the rest of the details. Maybe it's worth explicitly explaining this flow. The flow you've described sounds good though - it'd be useful for you to also briefly touch on how this could be improved in future to account for more complex degradations (e.g mitigating flappy backends by requiring multiple check passes before re-adding to the healthy pool)

File: rfd/0001-general-design.md [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+```
+This will adjust behavior to not pick the stale `[]*Route` if Update Routes started
+to mark the entries and replacing the slice reference.
+
+However, it will try to use the stale list once if it was fetched before Update Routes happened.
+
+### Rate Limiter
+
+###### Per-session: When TLS Handshake happened and session got authorization
+In this path we provide a Token bucket for each customer pool resource, which will
```
Comment (strideynet) [inline rfd/0001-general-design.md]: Is this limiting generally or per client identity ?

File: rfd/0001-general-design.md (hunk 2 above)
Comment (xdire) [reply rfd/0001-general-design.md:16]: Thanks for the hints; I will update it with an additional diagram. Currently found the following missing from the document:
- Least connection selector detailed flow (still slice reference explained, not sure if needed diagram for quite a simple loop)
- Health check flow diagram and actual flow explanation (or diagram)

I will try to add in 30m.

File: rfd/0001-general-design.md (hunk 16 above)
Comment (xdire) [reply rfd/0001-general-design.md]: As the configuration was not shared explicitly, it seems to be an open question; however, the following strategies can be applicable:
- per client pool (mentioned, as well normal behavior, as usually price is per LB )
- per client overall

In the case of per client account overall, it probably would be a Token Bucket with atomics attached to the customer map[clientUUID]*RateLimiter. However trying to imagine the case when client creates multiple forwarders and still accounted for the account and not per service :)

**Comment by xdire** [comment]
Document updated. Added health check scheduler section. Comments addressed.

**Comment by smallinsky** [comment]
> Document updated. Added health check scheduler section. Comments addressed.

Sorry,  but I can't find any updates related to following comment
https://github.com/xdire/xlb/pull/1#pullrequestreview-1985579728

**Comment by xdire** [comment]
> > Document updated. Added health check scheduler section. Comments addressed.
> 
> Sorry, but I can't find any updates related to following comment [#1 (review)](https://github.com/xdire/xlb/pull/1#pullrequestreview-1985579728)

@smallinsky Yes, I believe I assumed this part had some clarity from the previous high-level larger scope design document: https://github.com/xdire/xlb/blob/rfd/0002-grand-application-design/rfd/0002-grand-application-design.md#proto-specification.

However, as we moved to the reduced scope of the library, now the authorization layer is represented with: https://github.com/xdire/xlb/blob/rfd/0001-general-design/rfd/0001-general-design.md#security-authentication-and-credential-provision which states following after the scheme:

> Provided to Load Balancer with AddForwarderPool(pool ServicePool) method where ServicePool is the interface for the persistent configuration. ServicePool.GetIdentity() will be used to ensure CN match for authorization.

> This design will limit scope of Identity to simple selection of the whole set of the customer pool for such identity.

> As the future next steps certificate can be enriched with property like the OU=role and this will provide ability to have customer pools to routed for the authentication with additional precision.

Let me know if this helps or if it should be expanded even more.

**Review by smallinsky — APPROVED** [review]

**Review by strideynet — APPROVED** [review]
I'd appreciate the RFD including an example of the CLI/Config file, but, I'm happy enough with the rest of this so I'll approve to unblock you here.

**Comment by rosstimothy** [comment]
I can't seem to find any mention of how the least connection forwarding will be implemented. There are a few references to a strategy but there are no concrete details anywhere about how you plan to implement the only required strategy for the challenge.

**Comment by xdire** [comment]
> I can't seem to find any mention of how the least connection forwarding will be implemented. There are a few references to a strategy but there are no concrete details anywhere about how you plan to implement the only required strategy for the challenge.

@rosstimothy I believe this information is hiding in https://github.com/xdire/xlb/blob/rfd/0001-general-design/rfd/0001-general-design.md#strategy-next as the quote: `Provided strategy will use the current reference for provide the selection mechanics as a forward iteration loop.` and implies reference to the chart above like:

```
│         │     ┌───────┐       
│         │ ┌───┤*Route ├───┐   
│         │ │   └───────┘   │   
│         │ │┌─────────────┐│   
│         │ ││   Address   ││   
│         │ │└─────────────┘│   
│         │ │┌─────────────┐│   
│ ┌───────┼─▶│   Healthy   ││   
│ │       │ │└─────────────┘│   
│ │       │ │┌─────────────┐│   
│ ├───────┼─▶│ Connections ││   
│┌┴┐      │ │└─────────────┘│   
││A│      │ │┌─────────────┐│   
││T├──────┼─▶│   Active    ││   
││O│      │ │└─────────────┘│   
││M│      │ └───────────────┘   
││I│      │                     
││C│      └─────────────────────
│└┬┘                            
│ │  ┌──────────────┐           
│ │  │              │           
│ └──│   Strategy   │           
│    │              │              ┌────┴─────┐  
│    └──────────────┴────┐         │  Update  │  
│                      ╔═╩════╦────▶   Mutex  │  
│                      ║ Each ║    └────▲─────┘  
│                      ║Next()║         │        
│                      ╚══════╝                   ╚══════╝  
```
I was wondering if unwrapping more about the loop over the Routes with lookups into mentioned properties to calculate min() is worth any additional context. At the same time, I concentrated on unwrapping more complex topics of RR and HC in the way of an algorithm into design. Let me know if this topic still needs more to the table :) 

@strideynet thank you for unblocking, I need PoC for this design to be completed before the deadline :)

**Comment by rosstimothy** [comment]
> @rosstimothy I believe this information is hiding in https://github.com/xdire/xlb/blob/rfd/0001-general-design/rfd/0001-general-design.md#strategy-next as the quote: `Provided strategy will use the current reference for provide the selection mechanics as a forward iteration loop.` and implies reference to the chart above like:

Even after looking at the graph a few times and reading through the section that you linked I'm still not certain how you are proposing to implement the least connections algorithm.

**Comment by xdire** [comment]
> > @rosstimothy I believe this information is hiding in https://github.com/xdire/xlb/blob/rfd/0001-general-design/rfd/0001-general-design.md#strategy-next as the quote: `Provided strategy will use the current reference for provide the selection mechanics as a forward iteration loop.` and implies reference to the chart above like:
> 
> Even after looking at the graph a few times and reading through the section that you linked I'm still not certain how you are proposing to implement the least connections algorithm.

@rosstimothy I agree that from the standpoint of design, it mentions that algorithm is a loop but does not provide the details and there implication that someone need to read the words then look at the chart, and magically solution will appear in front of the eyes — which is definitely seems like a weak place in the document.

Would it help if the algorithm would be described in the following fashion?:
```sudo
// Function returns Next route available or nil
func Next() *Route
  min = maxInt
  selected = nil
  for each Route
      if route.Active && route.Healthy
          if route.Connections < min
              min = route.Connections
              selected = route
  return selected
```
As well Next() can return index number for further access, but I believe (at the moment it's all in imagination) mutex flow should be more blocking that way to assure there will be never OOB index access.

Let me know if we should have include this into the design and wait for the next round of approvals.

**Comment by rosstimothy** [comment]
> Would it help if the algorithm would be described in the following fashion?:
> 
> ```
> // Function returns Next route available or nil
> func Next() *Route
>   min = maxInt
>   selected = nil
>   for each Route
>       if route.Active && route.Healthy
>           if route.Connections < min
>               min = route.Connections
>               selected = route
>   return selected
> ```
> 
> As well Next() can return index number for further access, but I believe (at the moment it's all in imagination) mutex flow should be more blocking that way to assure there will be never OOB index access.
> 
> Let me know if we should have include this into the design and wait for the next round of approvals.

Yeah I think some description of your algorithm chosen to satisfy least connections would be helpful. Did you consider any alternate algorithms that don't require holding a lock and iterating over the entire route set?

**Comment by xdire** [comment]
> > Would it help if the algorithm would be described in the following fashion?:
> > ```
> > // Function returns Next route available or nil
> > func Next() *Route
> >   min = maxInt
> >   selected = nil
> >   for each Route
> >       if route.Active && route.Healthy
> >           if route.Connections < min
> >               min = route.Connections
> >               selected = route
> >   return selected
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As well Next() can return index number for further access, but I believe (at the moment it's all in imagination) mutex flow should be more blocking that way to assure there will be never OOB index access.
> > Let me know if we should have include this into the design and wait for the next round of approvals.
> 
> Yeah I think some description of your algorithm chosen to satisfy least connections would be helpful. Did you consider any alternate algorithms that don't require holding a lock and iterating over the entire route set?
@rosstimothy 
All right, I'll update the RFD.

Regarding the algorithm and locks, the mutex is chosen just for the reasons of the list to be dynamically updated:
#### Lock and update conundrum
1) Lock there is just to get the active pointer to the list in a synchronized fashion. The iteration comes lock-free, but the list might be stale for one iteration.
2) This will lead mutex to race; however, it should be quick to resolve
3) If we don't support dynamic route updates, e.g. LB will shut down and be reinstated with a new config — then we no need a lock and can just read
#### Choices of loop vs something else:
###### Factors
- the common size of the routes is unknown, but let's assume we balancing for 3-9 services or 1000 services
- smart methods to pop up the next route would give us NlogN (PQ, Sort)
- as for benchmark goes in most cases, I observed benchmarks comparing the time execution for PQ vs. Sort: https://stackoverflow.com/a/3759564 it seems for them being both NlogN it's not apples to apples.
- however, while validating the approach, I observed most implementations in libraries for the least connections are just forward loops, and it seems that this possibly derived from the real cases of how much traffic will come through the single LB, for example:
  - let's say we try to serve 1000 machines
  - assuming this kind of load, that 1000 machines required to sustain it is either slow computational heavy problems and traffic is slow (1000rps), or rapid traffic spreading across 1000 services, which could be 1000rps * 1000services = 1M connections and the single instance will not sustain this, meaning it'll be more than one LB working out this algorithm.

###### Conclusions
After a few cycles of the research, I decided not to try to reinvent the wheel at this time, as well in this paper: https://www.irjet.net/archives/V5/i3/IRJET-V5I3455.pdf it's also a forward loop for the least connection part.
